hail/src/main/scala/is/hail/expr/AnnotationImpex.scala

```diff
@@ -25,17 +26,17 @@ object SparkAnnotationImpex {
     }.mkString
   }
 
-  def importType(t: DataType): Type = t match {
-    case BooleanType => TBoolean()
-    case IntegerType => TInt32()
-    case LongType => TInt64()
-    case FloatType => TFloat32()
-    case DoubleType => TFloat64()
-    case StringType => TString()
-    case BinaryType => TBinary()
-    case ArrayType(elementType, containsNull) => TArray(importType(elementType).setRequired(!containsNull))
+  def importType(t: DataType): PType = t match {
```

Comment:
Want to check: are there any stricter requiredeness rules we can apply here? One thought: if ArrayType.containsNull is false, one interpretation. is that the array itself cannot be null (the other is that the outer array can be missing, but when present contains no nulls).

Reply:
nope, this is as good a we can do.

Reply:
What are the semantics of containsNull?

Documentation opposite of detailed: https://spark.apache.org/docs/2.1.1/api/java/org/apache/spark/sql/types/ArrayType.html#containsNull()

Reply:
containsNull means that the elements can be null. So !containsNull means that the element should be required.

Reply:
that's what we do here

Reply:
Yeah, I was curious about whether this also gave information about the Array itself (the container). 

So now I understand: the type can always be missing, but children of a type can be  never-null